<a href="https://colab.research.google.com/github/agmalaga2020/PolicySpace2_Spanish_data/blob/main/tama%C3%B1o_medio_hogares_ccaa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INICIO

- 1º buscamos fuente a explorar
https://uoc0-my.sharepoint.com/:x:/g/personal/agmalaga_uoc_edu/EVrtEuZkGqxMsXz3ENcmP8sB8w9sxyyFWBX3pDP282BDaQ?e=hZX8jz

_______________________________________________________________

- 2º `CARGAMOS` DATAFRAME ORIGINAL del proyecto **github**
https://github.com/agmalaga2020/PolicySpace2-agmalaga/tree/master/input

_______________________________________________________________
- 3º Realizamos analisis exploratorio en google colab.
_______________________________________________________________

- 4º descargamos datos y creamos informe.
_______________________________________________________________

- 5º actualizar LINK de la columna `NOTEBOOK`
https://uoc0-my.sharepoint.com/:x:/g/personal/agmalaga_uoc_edu/EVrtEuZkGqxMsXz3ENcmP8sB8w9sxyyFWBX3pDP282BDaQ?e=hZX8jz

_______________________________________________________________

ESQUEMA DE PROCESO

https://whiteboard.office.com/me/whiteboards/p/c3BvOmh0dHBzOi8vdW9jMC1teS5zaGFyZXBvaW50LmNvbS9wZXJzb25hbC9hZ21hbGFnYV91b2NfZWR1/b!r4X97qYbxUyQGQNSpn-bfrNi376yDThFsrR2Kul1sxTx_nU34TfBSpdz-6pNWzvq/01E765GIPQJMS23ADWVBEID5XGUDPDSQNZ?source=applauncher&auth_upn=agmalaga%40uoc.edu

### average_num_members_families_2010.csv

'https://raw.githubusercontent.com/BAFurtado/PolicySpace2/refs/heads/master/input/average_num_members_families_2010.csv'


In [3]:
import requests
import pandas as pd
import io

# Update the URL to point to the raw CSV data on GitHub
url = 'https://raw.githubusercontent.com/BAFurtado/PolicySpace2/refs/heads/master/input/average_num_members_families_2010.csv'
s = requests.get(url).content

# Read the CSV data using ';' as the separator
df_original = pd.read_csv(io.StringIO(s.decode('utf-8')), sep=';')

from google.colab import files
df_original.to_csv('df_original.csv', encoding = 'utf-8-sig')
files.download('df_original.csv')

print(df_original.head())

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

               AREAP,avg_num_people
0  1302603005001,4.0948936170212775
1  1302603005002,4.1957971014492745
2  1302603005003,3.9286440677966112
3   1302603005004,4.084615384615383
4   1302603005005,3.863297872340424


# ESPAÑA



Perfecto. Vamos con el análisis de los datos del proyecto original respecto al archivo `average_num_members_families_2010.csv` y su posible adaptación al contexto español. Aquí tienes el esquema siguiendo tu índice:

---

### 📂 Archivo: `average_num_members_families_2010.csv`

🔍 **¿A qué hace referencia exactamente?**  
Este archivo representa el tamaño medio de los hogares, es decir, el número promedio de miembros por familia que residen en viviendas particulares. Se trata de una de las variables base que utiliza *PolicySpace2* para simular dinámicas demográficas y de demanda de vivienda. Es una medida importante porque afecta a la formación de nuevos hogares, la presión sobre el mercado inmobiliario y la distribución del gasto familiar.

🌍 **¿Cómo están definidos geográficamente?**  
En *PolicySpace2*, esta variable se define para cada una de las regiones metropolitanas simuladas en Brasil (46 en total). Los datos provienen del censo brasileño de 2010, y se utiliza el valor promedio de tamaño de familia para iniciar la generación de agentes (households) en cada región.

🛠 **¿Cómo adaptarlo al contexto Español?**  
Nuestros datos del INE permiten replicar esta variable fácilmente. Ya que en España contamos con estadísticas del tamaño medio de los hogares por **comunidad autónoma**, podemos utilizarlas como proxy directa para inicializar los agentes en nuestro modelo. Dado que *PolicySpace2* trabaja con regiones metropolitanas, y nuestros datos están agregados por comunidades, una opción es:

- Asignar directamente el valor de la comunidad autónoma a todas las provincias dentro de ella.
- O bien, si se quieren afinar más las simulaciones, desagregar por provincias o áreas funcionales urbanas si se dispone de esos datos más desglosados.

🧠 **Conclusión para el proyecto**  
Podemos adaptar perfectamente esta variable al modelo español sin perder riqueza descriptiva. Basta con usar el valor del tamaño medio de hogar para cada comunidad autónoma según el INE (año 2021 o más reciente), como punto de partida para la simulación de hogares. Esta variable es fundamental porque determina el ritmo de creación de nuevos hogares, un factor clave en la presión sobre el mercado inmobiliario y la demanda de vivienda social o ayudas al alquiler.



## Tamaño medio de los hogares de personas residentes en viviendas familiares por fecha

- Series desde 2021


https://www.ine.es/jaxiT3/Tabla.htm?t=60132

### extraemos las tablas

In [9]:
import requests
import pandas as pd

# Código de la tabla
table_code = "60132"
url_csv = f"https://servicios.ine.es/wstempus/csv/ES/DATOS_TABLA/{table_code}?nult=999"

response = requests.get(url_csv)

if response.status_code == 200:
    csv_data = response.content.decode('utf-8')

    # Guardar el archivo CSV con un nombre adecuado usando el código de la tabla
    file_name = f"tabla_{table_code}.csv"
    with open(file_name, "w", encoding="utf-8") as f:
        f.write(csv_data)
    print(f"Archivo CSV guardado como '{file_name}'")
else:
    print("Error en la petición: ❌", response.status_code)

# Leer el CSV
df = pd.read_csv(file_name, sep='\t', encoding='utf-8')

# Mostrar las primeras filas
print(df.head())
print("Tabla actualizada conseguida ✅")

try:
    df.to_csv(f'/content/{table_code}.csv', index=False)
    print(f"✅ Copia de df guardada en '/content/{table_code}.csv'")
    print("Se ha guardado una copia del DataFrame df.")
except NameError:
    print("⚠️ df o table_code no están definidos. No se puede guardar la copia.")

Archivo CSV guardado como 'tabla_60132.csv'
  Comunidades y ciudades autónomas               Periodo Total
0                   Total Nacional    1 de enero de 2025  2,51
1                   Total Nacional  1 de octubre de 2024  2,51
2                   Total Nacional    1 de julio de 2024  2,50
3                   Total Nacional    1 de abril de 2024  2,50
4                   Total Nacional    1 de enero de 2024  2,50
Tabla actualizada conseguida ✅
✅ Copia de df guardada en '/content/60132.csv'
Se ha guardado una copia del DataFrame df.


In [10]:
# Usar la función de exploración definida para ver detalles de las columnas
def print_unique_values_and_len(df):
    """Imprime valores únicos, número de NaN y tipo por columna."""
    for column in df.columns:
        unique_values = df[column].unique()
        nan_count = df[column].isnull().sum()
        dtype = df[column].dtype
        print(f"🧩 Columna: {column}")
        print(f"- Valores únicos: {unique_values}")
        print(f"- Total únicos: {len(unique_values)}")
        print(f"- NaNs: {nan_count}")
        print(f"- Tipo: {dtype}\n")

print_unique_values_and_len(df)


🧩 Columna: Comunidades y ciudades autónomas
- Valores únicos: ['Total Nacional' 'Andalucía' 'Aragón' 'Asturias, Principado de'
 'Balears, Illes' 'Canarias' 'Cantabria' 'Castilla y León'
 'Castilla - La Mancha' 'Cataluña' 'Comunitat Valenciana' 'Extremadura'
 'Galicia' 'Madrid, Comunidad de' 'Murcia, Región de'
 'Navarra, Comunidad Foral de' 'País Vasco' 'Rioja, La' 'Ceuta' 'Melilla']
- Total únicos: 20
- NaNs: 0
- Tipo: object

🧩 Columna: Periodo
- Valores únicos: ['1 de enero de 2025' '1 de octubre de 2024' '1 de julio de 2024'
 '1 de abril de 2024' '1 de enero de 2024' '1 de octubre de 2023'
 '1 de julio de 2023' '1 de abril de 2023' '1 de enero de 2023'
 '1 de octubre de 2022' '1 de julio de 2022' '1 de abril de 2022'
 '1 de enero de 2022' '1 de octubre de 2021' '1 de julio de 2021'
 '1 de abril de 2021' '1 de enero de 2021']
- Total únicos: 17
- NaNs: 0
- Tipo: object

🧩 Columna: Total
- Valores únicos: ['2,51' '2,50' '2,52' '2,53' '2,54' '2,55' '2,56' '2,57' '2,58' '2,59'
 '2,60' 

### datos comunidades

In [11]:
import pandas as pd
import warnings

# Suprime los warnings
warnings.filterwarnings('ignore')

# URL y lectura de la primera tabla
url = 'https://www.ine.es/daco/daco42/codmun/cod_ccaa_provincia.htm'
df_temp = pd.read_html(url, encoding='ISO-8859-1')[0]
df_temp.columns = ['CODAUTO', 'Comunidad Autónoma', 'CPRO', 'Provincia']

# Filtra para eliminar "Ciudades Autónomas" y convierte CODAUTO a numérico
df_temp = df_temp[df_temp['Comunidad Autónoma'] != 'Ciudades Autónomas']
df_temp['CODAUTO'] = pd.to_numeric(df_temp['CODAUTO'], errors='coerce')
df_temp = df_temp.dropna(subset=['CODAUTO'])

# Guarda el DataFrame en CSV y muestra las primeras filas
df_temp.to_csv('codigos_ccaa_provincias.csv', index=False, encoding='utf-8')
print(df_temp.head())

# Crea y muestra el DataFrame de equivalencias de comunidades
equivalencias_comunidades = df_temp[['CODAUTO', 'Comunidad Autónoma']].drop_duplicates('CODAUTO')
print(equivalencias_comunidades)


   CODAUTO Comunidad Autónoma  CPRO Provincia
0      1.0          Andalucía   4.0   Almería
1      1.0          Andalucía  11.0     Cádiz
2      1.0          Andalucía  14.0   Córdoba
3      1.0          Andalucía  18.0   Granada
4      1.0          Andalucía  21.0    Huelva
    CODAUTO           Comunidad Autónoma
0       1.0                    Andalucía
8       2.0                       Aragón
11      3.0      Asturias, Principado de
12      4.0               Balears, Illes
13      5.0                     Canarias
15      6.0                    Cantabria
16      7.0              Castilla y León
25      8.0           Castilla-La Mancha
30      9.0                     Cataluña
34     10.0         Comunitat Valenciana
37     11.0                  Extremadura
39     12.0                      Galicia
43     13.0         Madrid, Comunidad de
44     14.0            Murcia, Región de
45     15.0  Navarra, Comunidad Foral de
46     16.0                   País Vasco
49     17.0                


- ✅ `Comunidades y ciudades autónomas` → `comunidad_name`
- ✅ `Periodo` separado en `año` y `mes` (formato numérico de mes)
- ✅ `Total` convertido a tipo `float` como `total`
- ✅ Añadida columna `comunidad_code` a partir de `equivalencias_comunidades.csv`



In [16]:
# Copia del DataFrame original para limpieza
df_clean = df.copy()

# Renombrar columna
df_clean.rename(columns={'Comunidades y ciudades autónomas': 'comunidad_name'}, inplace=True)

# Extraer año y mes de la columna 'Periodo'
meses_dict = {
    'enero': '01', 'febrero': '02', 'marzo': '03', 'abril': '04',
    'mayo': '05', 'junio': '06', 'julio': '07', 'agosto': '08',
    'septiembre': '09', 'octubre': '10', 'noviembre': '11', 'diciembre': '12'
}

# Crear nuevas columnas 'año' y 'mes'
df_clean['mes'] = df_clean['Periodo'].str.extract(r'de (\w+) de')[0].str.lower()
df_clean['mes'] = df_clean['mes'].map(meses_dict)
df_clean['año'] = df_clean['Periodo'].str.extract(r'de (\d{4})')[0].astype(int)

# Convertir la columna 'Total' a formato numérico (cambiando coma por punto)
df_clean['total'] = df_clean['Total'].str.replace(',', '.').astype(float)

# Unir con dataset de equivalencias para obtener comunidad_code
df_clean = df_clean.merge(equivalencias_comunidades, left_on='comunidad_name', right_on='Comunidad Autónoma', how='left')
df_clean.rename(columns={'CODAUTO': 'comunidad_code'}, inplace=True)

# Eliminar columnas no necesarias
df_clean.drop(columns=['Periodo', 'Total', 'Comunidad Autónoma'], inplace=True)

# Eliminar filas donde comunidad_name es 'Total Nacional'
df_clean = df_clean[df_clean['comunidad_name'] != 'Total Nacional']

# Mostrar dimensiones después del filtrado
v.shape



(323, 5)

In [19]:
print_unique_values_and_len(df_clean)

🧩 Columna: comunidad_name
- Valores únicos: ['Andalucía' 'Aragón' 'Asturias, Principado de' 'Balears, Illes'
 'Canarias' 'Cantabria' 'Castilla y León' 'Castilla - La Mancha'
 'Cataluña' 'Comunitat Valenciana' 'Extremadura' 'Galicia'
 'Madrid, Comunidad de' 'Murcia, Región de' 'Navarra, Comunidad Foral de'
 'País Vasco' 'Rioja, La' 'Ceuta' 'Melilla']
- Total únicos: 19
- NaNs: 0
- Tipo: object

🧩 Columna: mes
- Valores únicos: ['01' '10' '07' '04']
- Total únicos: 4
- NaNs: 0
- Tipo: object

🧩 Columna: año
- Valores únicos: [2025 2024 2023 2022 2021]
- Total únicos: 5
- NaNs: 0
- Tipo: int64

🧩 Columna: total
- Valores únicos: [2.55 2.54 2.56 2.57 2.58 2.59 2.6  2.4  2.39 2.42 2.44 2.22 2.21 2.23
 2.24 2.25 2.66 2.65 2.64 2.63 2.61 2.62 2.41 2.43 2.26 2.27 2.28 2.29
 2.3  2.52 2.51 2.53 2.48 2.47 2.46 2.49 2.37 2.36 2.38 2.45 2.75 2.74
 2.76 2.77 2.78 2.79 2.35 2.34 3.14 3.15 3.16 3.17 3.18 3.19 3.2  3.21
 3.23 3.25 3.27 3.28 3.31 3.34]
- Total únicos: 62
- NaNs: 0
- Tipo: float64

🧩 Co

In [20]:
equivalencias_comunidades[equivalencias_comunidades['CODAUTO'] == 8]

CODAUTO  Comunidad Autónoma
25      8.0  Castilla-La Mancha

In [23]:
df_nan_codauto = df_clean[df_clean['comunidad_code'].isna()]
df_nan_codauto.comunidad_name.unique()

array(['Castilla - La Mancha'], dtype=object)

In [24]:
df_clean

comunidad_name mes   año  total  comunidad_code
17       Andalucía  01  2025   2.55             1.0
18       Andalucía  10  2024   2.55             1.0
19       Andalucía  07  2024   2.55             1.0
20       Andalucía  04  2024   2.55             1.0
21       Andalucía  01  2024   2.54             1.0
..             ...  ..   ...    ...             ...
335        Melilla  01  2022   3.25            19.0
336        Melilla  10  2021   3.27            19.0
337        Melilla  07  2021   3.28            19.0
338        Melilla  04  2021   3.31            19.0
339        Melilla  01  2021   3.34            19.0

[323 rows x 5 columns]

In [34]:
# Sustituir valores NaN en 'comunidad_code' por 8 donde 'comunidad_name' sea 'Castilla - La Mancha'
df_clean.loc[(df_clean['comunidad_name'] == 'Castilla-La Mancha') & (df_clean['comunidad_code'].isna()), 'comunidad_code'] = 8

df_nan_codauto = df_clean[df_clean['comunidad_code'].isna()]
df_nan_codauto.comunidad_name.unique()

array([], dtype=object)

In [35]:
print(df_clean.isna().sum().sum())

0


In [44]:
df_andalucia_2024 = df_clean[(df_clean['comunidad_name'] == 'Andalucía') & (df_clean['año'] == 2024)]
print("control 2025 Andalucía", df_andalucia_2024.total.mean())
df_andalucia_2024

control 2025 Andalucía 2.5475


comunidad_name mes   año  total  comunidad_code
18      Andalucía  10  2024   2.55             1.0
19      Andalucía  07  2024   2.55             1.0
20      Andalucía  04  2024   2.55             1.0
21      Andalucía  01  2024   2.54             1.0

In [37]:
# Agrupar por comunidad_code y año, y calcular la media del total
df_media_anual = df_clean.groupby(['comunidad_code', 'año'], as_index=False)['total'].mean()
df_media_anual

comunidad_code   año   total
0              1.0  2021  2.5875
1              1.0  2022  2.5650
2              1.0  2023  2.5500
3              1.0  2024  2.5475
4              1.0  2025  2.5500
..             ...   ...     ...
90            19.0  2021  3.3000
91            19.0  2022  3.2250
92            19.0  2023  3.2000
93            19.0  2024  3.1950
94            19.0  2025  3.2000

[95 rows x 3 columns]

### guardamos df tamaño_medio_hogares_ccaa_completo.csv



In [46]:
df_media_anual.to_csv("/content/tamaño_medio_hogares_ccaa_completo.csv", index=False)
from google.colab import files
files.download("/content/tamaño_medio_hogares_ccaa_completo.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [47]:
df_media_anual

comunidad_code   año   total
0              1.0  2021  2.5875
1              1.0  2022  2.5650
2              1.0  2023  2.5500
3              1.0  2024  2.5475
4              1.0  2025  2.5500
..             ...   ...     ...
90            19.0  2021  3.3000
91            19.0  2022  3.2250
92            19.0  2023  3.2000
93            19.0  2024  3.1950
94            19.0  2025  3.2000

[95 rows x 3 columns]

In [52]:
import pandas as pd
import os

# Create the directory if it doesn't exist
output_dir = "/content/data final"
os.makedirs(output_dir, exist_ok=True)

# Create a dictionary to store the dataframes by year
dataframes_por_año = {}

# Group by year and save each as CSV
for año, df_año in df_media_anual.groupby("año"):
    nombre = f"tamaño_medio_hogares_ccaa_{año}"
    dataframes_por_año[nombre] = df_año
    df_año.to_csv(f"{output_dir}/{nombre}.csv", index=False)

# Show the names of the generated files
print(list(dataframes_por_año.keys()))


[]


In [55]:
import os

# Create the directory if it doesn't exist
output_dir = "/content/data final"
os.makedirs(output_dir, exist_ok=True)

# Assuming df_media_anual is already defined from the previous code

# Guardar nuevamente los archivos por año pero eliminando la columna 'año'
for año, df_año in df_media_anual.groupby("año"):
    nombre = f"tamaño_medio_hogares_ccaa_{año}"
    df_año_sin_anio = df_año.drop(columns=["año"])
    df_año_sin_anio.to_csv(f"{output_dir}/{nombre}.csv", index=False) # Use f-string for path

#Instead, generate the filenames based on the years in the dataframe
file_names = [f"tamaño_medio_hogares_ccaa_{year}.csv" for year in df_media_anual['año'].unique()]
file_names


['tamaño_medio_hogares_ccaa_2021.csv',
 'tamaño_medio_hogares_ccaa_2022.csv',
 'tamaño_medio_hogares_ccaa_2023.csv',
 'tamaño_medio_hogares_ccaa_2024.csv',
 'tamaño_medio_hogares_ccaa_2025.csv']

## 📊 Informe de Procesamiento y Transformación de Datos



### 🧼 Proceso de Limpieza

| Paso                     | Descripción                                                                 |
|--------------------------|-----------------------------------------------------------------------------|
| Carga inicial            | Se cargó el archivo CSV `tabla_60132.csv`                                   |
| Filtrado                 | Se eliminaron las filas con `comunidad_name = "Total Nacional"`             |
| Normalización            | Renombramos columnas para seguir la convención `snake_case`                |
| Conversión numérica      | La columna `Total` se transformó de string con coma decimal a `float`       |
| Separación de columnas   | La columna `Periodo` se separó en `año` y `mes`                             |
| Unión con equivalencias  | Se añadió `comunidad_code` a partir del archivo `equivalencias_comunidades`|
| Agrupación               | Se calculó la media anual del tamaño de hogar por comunidad y año           |
| División por años        | Se generaron 5 archivos CSV, uno por año, sin la columna `año`              |

### 🧪 Estadísticas de DataFrames

| DataFrame                           | Filas | Columnas |
|------------------------------------|-------|----------|
| df_clean | 323 | 5 |
| df_media_anual | 95 | 3 |
| tamaño_medio_hogares_ccaa_2021 | 19 | 3 |
| tamaño_medio_hogares_ccaa_2022 | 19 | 3 |
| tamaño_medio_hogares_ccaa_2023 | 19 | 3 |
| tamaño_medio_hogares_ccaa_2024 | 19 | 3 |
| tamaño_medio_hogares_ccaa_2025 | 19 | 3 |


---

### 🧠 Conclusión

Este conjunto de datos proporciona una base sólida y estandarizada para incorporar el tamaño medio de los hogares como una variable demográfica clave en simulaciones de modelos basados en agentes como *PolicySpace2*. Al integrarlo, se podrán analizar con mayor precisión las dinámicas sociales y económicas relacionadas con el bienestar, la desigualdad y la presión sobre infraestructuras públicas a nivel regional en España.

---

